File: backend/src/main/java/com/just/donate/flow/DonationQueue.java
package com.just.donate.flow;

import java.math.BigDecimal;
import java.util.List;

class DonationQueue {

    private final ReservableQueue<DonationPart, BigDecimal, Account> donations;
    private final BigDecimal negativeBalance = BigDecimal.ZERO;

    public DonationQueue(Account account) {
        this.donations = new ReservableQueue<>(account);
    }

    public void add(DonationPart donation) {
        donations.add(donation);
    }
    
    public void addAll(List<DonationPart> donations) {
        donations.forEach(this::add);
    }

    public BigDecimal totalBalance() {
        return donations.getQueue().map(r -> r.getValue().getAmount()).foldLeft(BigDecimal.ZERO, BigDecimal::add);
    }
    
    public ReservableQueue<DonationPart, BigDecimal, Account> getDonationsQueue() {
        return donations;
    }
}

File: backend/src/main/scala/com/just/donate/models/DonationQueue.scala
package com.just.donate.models

import com.just.donate.utils.Money
import com.just.donate.utils.structs.ReservableQueue

import scala.math.Ordered.orderingToOrdered
case class DonationQueue(
  context: String,
  donationQueue: ReservableQueue,
  negativeBalance: Money = Money.ZERO
):

  def add(donation: DonationPart): DonationQueue =
    copy(donationQueue = donationQueue.add(donation))

  def addAll(donations: Seq[DonationPart]): DonationQueue =
    copy(donationQueue = donations.foldLeft(donationQueue)(_.add(_)))

  def totalBalance: Money = donationQueue.queue.map(_.value.amount).sum

  def pull(amount: Money, limit: Option[Int] = None): (Seq[DonationPart], Option[Money], DonationQueue) =
    val (donations, remaining, newQueue) = donationQueue.pollUnreserved(amount, limit)
    (donations, remaining, copy(donationQueue = newQueue))
